### PR TITLE
Feature & Refactor: read and write Hexx(R) in CSR format

### PR DESCRIPTION
**Pull request by maki49**
### Background
Currently, EXX-NSCF suffers from cereal's bug in reading Hexx(R) (related issues: #2235 #3323), which forces the users to turn to PyATB.

### What's changed
- refactor (unify) the CSR writers
- read and write Hexx(R) in CSR format to bypass cereal
- add a parameter "reduce" to CSR writers: if false, each process will write its data respectively to its file.

### CODE REVIEW

File: source/module_ri/Exx_LRI_interface.h [hunk 1]
```diff
@@ -19,15 +19,24 @@ template<typename T, typename Tdata>
...
     /// @param exx_ptr
     Exx_LRI_Interface(std::shared_ptr<Exx_LRI<Tdata>> exx_ptr) : exx_ptr(exx_ptr) {}
     Exx_LRI_Interface() = delete;
 
     void write_Hexxs(const std::string &file_name) const;
     void read_Hexxs(const std::string& file_name);
-    
-    using TAC = std::pair<int, std::array<int, 3>>;
+
+    void write_Hexxs(const std::string& file_name, const UnitCell& ucell) const;
```
Comment (dyzheng) [inline source/module_ri/Exx_LRI_interface.h]: Can you add some annotations for the difference between these two new functions and old two functions above?

File: source/module_ri/Exx_LRI_interface.h [hunk 2]
```diff
@@ -19,15 +19,24 @@ template<typename T, typename Tdata>
...
+    void read_Hexxs(const std::string& file_name, const UnitCell& ucell);
+
+    std::map<Abfs::Vector3_Order<int>, std::map<size_t, std::map<size_t, Tdata>>>
```
Comment (dyzheng) [inline source/module_ri/Exx_LRI_interface.h]: There is a function with complex return, you'd better add annotations to describe the function and the returned value.

**Comment by PeizeLin** [comment]
All HexxR have been reduced to only one process. Will this lead to memory crashes in large-scale computing?

**Comment by maki49** [comment]
> All HexxR have been reduced to only one process. Will this lead to memory crashes in large-scale computing?

`exx_lri.Hexxs` is already a global matrix (with the same value on all the processors) no need for extra reduction. 
What I did is simply that: if `DRANK==0`, convert to CSR and write; else, do nothing.

**Comment by PeizeLin** [comment]
> > All HexxR have been reduced to only one process. Will this lead to memory crashes in large-scale computing?
> 
> `exx_lri.Hexxs` is already a global matrix (with the same value on all the processors) no need for extra reduction. What I did is simply that: if `DRANK==0`, convert to CSR and write; else, do nothing.

For each process, only part of Hexx is stored in memory.
Although the same Hexx[I][J] may appear repeatedly in multiple processes, it does not mean that each process contains all Hexx.
For large systems, the memory requirement to store the global Hamiltonian matrices is unacceptable.

**Comment by maki49** [comment]
> For each process, only part of Hexx is stored in memory. Although the same Hexx[I][J] may appear repeatedly in multiple processes, it does not mean that each process contains all Hexx. For large systems, the memory requirement to store the global Hamiltonian matrices is unacceptable.

So there's a waste of memory in Hexx to be improved.  Maybe I can have a try in another PR. 
This PR is to circumvent cereal's bug, as long as to "convert to CSR and write" in every processor can achieve this goal, right?

**Comment by PeizeLin** [comment]
> > For each process, only part of Hexx is stored in memory. Although the same Hexx[I][J] may appear repeatedly in multiple processes, it does not mean that each process contains all Hexx. For large systems, the memory requirement to store the global Hamiltonian matrices is unacceptable.
> 
> So there's a waste of memory in EXX Hexx to be improved. Maybe I can have a try in another PR. This PR is to circumvent cereal's bug, as long as to "convert to CSR and write" in every processor can achieve this goal, right?

right

File: source/module_ri/Exx_LRI_interface.h (hunk 1 above)
Comment (PeizeLin) [reply source/module_ri/Exx_LRI_interface.h]: @maki49 Maybe the function names can be changed to `write_Hexxs_cereal()` and `write_Hexxs_csr()` explicitly?

File: source/module_esolver/esolver_ks_lcao.cpp [hunk 3]
```diff
@@ -895,11 +895,11 @@ namespace ModuleESolver
 #ifdef __EXX
     if (GlobalC::exx_info.info_global.cal_exx) // Peize Lin add if 2022.11.14
     {
-        const std::string file_name_exx = GlobalV::global_out_dir + "HexxR_" + std::to_string(GlobalV::MY_RANK);
+        const std::string file_name_exx = GlobalV::global_out_dir + "HexxR" + std::to_string(GlobalV::MY_RANK);
```
Comment (PeizeLin) [inline source/module_esolver/esolver_ks_lcao.cpp:1027]: @maki49 Maybe the filename can also be changed to "HexxR_0_csr", avoiding users to read in old binary cereal files with new CSR code.

**Review by PeizeLin — APPROVED** [review]

**Review by PeizeLin — APPROVED** [review]

**Review by dyzheng — APPROVED** [review]

**Pull request merged**